tracer/src/Datadog.Trace/Iast/EvidenceConverter.cs

```diff
@@ -95,145 +105,398 @@ private void ToJsonTaintedValue(JsonWriter writer, string value, Range[] ranges)
+        {
+            var parts = new List<ValuePart>();
+            if (Value != null)
+            {
+                if (Source != null && Source.IsSensitive)
                 {
-                    nextTainted.Value.Source?.MarkAsRedacted();
-                    nextSensitive = Poll(sensitive);
+                    // redact the full tainted value as the source is sensitive (password, certificate, ...)
+                    AddValuePart(0, Value.Length, true, parts);
                 }
-
-                // If the tainted range intersects sensitive ranges, we redact the tainted range source and split the sensitive range
-                if (nextSensitive != null && nextSensitive.Value.Intersects(nextTainted.Value))
+                else if (SensitiveRanges != null)
                 {
-                    nextTainted.Value.Source?.MarkAsRedacted();
-                    var rest = nextSensitive.Value.Remove(nextTainted.Value);
-                    nextSensitive = rest.Count > 0 ? rest[0] : null;
+                    // redact only sensitive parts
+                    int index = 0;
+                    foreach (var sensitive in SensitiveRanges)
+                    {
+                        var start = sensitive.Start;
+                        var end = sensitive.Start + sensitive.Length;
+                        // append previous tainted chunk (if any)
+                        AddValuePart(index, start, false, parts);
+                        // append current sensitive tainted chunk
+                        AddValuePart(start, end, true, parts);
+                        index = end;
+                    }
+
+                    // append last tainted chunk (if any)
+                    AddValuePart(index, Value.Length, false, parts);
                 }
+            }
 
-                // We write the range value redacted or not depensing on the source status
-                if (nextTainted.Value.Source != null && nextTainted.Value.Source.IsRedacted())
+            return parts;
+        }
+
+        private void AddValuePart(int start, int end, bool redact, List<ValuePart> valueParts)
+        {
+            if (start < end)
+            {
+                var chunk = Value!.Substring(start, end - start);
+                if (!redact)
                 {
-                    WriteRedactedValuePart(writer, nextTainted.Value);
+                    // append the value
+                    valueParts.Add(new ValuePart(chunk, Source));
                 }
                 else
                 {
-                    WriteValuePart(writer, Substring(value, nextTainted.Value), nextTainted);
+                    var length = chunk.Length;
+                    var matching = Source!.Value!.IndexOf(chunk);
+                    if (matching >= 0)
+                    {
+                        // if matches append the matching part from the redacted value
+                        var pattern = Source!.RedactedValue!.Substring(matching, length);
```

Comment:
same for `Source` and `RedactedValue`, how do we know they're not going to be null?
if we are sure, maybe add a comment as why?

Reply:
Fixed in 97ba51b49cacea48d20b4965b4c7856615efcb63